File: mobile/lib/core/models/venue.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'venue.freezed.dart';
part 'venue.g.dart';

@freezed
abstract class Venue with _$Venue {
  const factory Venue({
    required String id,
    required String name,
    required String address,
    required String city,
    String? district,
    required double latitude,
    required double longitude,
    @JsonKey(name: 'google_place_id') String? googlePlaceId,
    String? notes,
    @Default('pending') String status,
    @JsonKey(name: 'rejection_note') String? rejectionNote,
    @JsonKey(name: 'added_by') required String addedBy,
    @JsonKey(name: 'added_by_name') String? addedByName,
    @JsonKey(name: 'approved_by') String? approvedBy,
    @JsonKey(name: 'verified_at') DateTime? verifiedAt,
    @JsonKey(name: 'verification_due_at') DateTime? verificationDueAt,
    @JsonKey(name: 'food_halal_mode') @Default('selected') String foodHalalMode,
    @JsonKey(name: 'excluded_products') @Default([]) List<String> excludedProducts,
    @Default([]) List<HalalCriteria> criteria,
    @Default([]) List<VenuePhoto> photos,
    @JsonKey(name: 'food_items') @Default([]) List<FoodItem> foodItems,
    @JsonKey(name: 'average_rating') double? avgRating,
    @JsonKey(name: 'review_count') @Default(0) int reviewCount,
    double? distance,
    @JsonKey(name: 'categories_str') String? categoriesStr,
    @JsonKey(name: 'created_at') DateTime? createdAt,
    @JsonKey(name: 'updated_at') DateTime? updatedAt,
  }) = _Venue;

  const Venue._();

  factory Venue.fromJson(Map<String, dynamic> json) => _$VenueFromJson(json);

  bool get isApproved => status == 'approved';
  bool get isPending => status == 'pending';
  bool get isRejected => status == 'rejected';
}

@freezed
abstract class HalalCriteria with _$HalalCriteria {
  const factory HalalCriteria({
    required int id,
    required String key,
    @JsonKey(name: 'label_tr') required String labelTr,
    @JsonKey(name: 'label_en') required String labelEn,
    @JsonKey(name: 'description_tr') String? descriptionTr,
    @JsonKey(name: 'description_en') String? descriptionEn,
  }) = _HalalCriteria;

  factory HalalCriteria.fromJson(Map<String, dynamic> json) =>
      _$HalalCriteriaFromJson(json);
}

@freezed
abstract class VenuePhoto with _$VenuePhoto {
  const factory VenuePhoto({
    required String id,
    @JsonKey(name: 'venue_id') required String venueId,
    required String url,
    @JsonKey(name: 'uploaded_by') required String uploadedBy,
    @JsonKey(name: 'is_primary') @Default(false) bool isPrimary,
    @JsonKey(name: 'created_at') DateTime? createdAt,
  }) = _VenuePhoto;

  factory VenuePhoto.fromJson(Map<String, dynamic> json) =>
      _$VenuePhotoFromJson(json);
}

@freezed
abstract class FoodCategory with _$FoodCategory {
  const factory FoodCategory({
    required int id,
    required String key,
    @JsonKey(name: 'label_tr') required String labelTr,
    @JsonKey(name: 'label_en') required String labelEn,
    @Default([]) List<FoodItem> items,
  }) = _FoodCategory;

  factory FoodCategory.fromJson(Map<String, dynamic> json) =>
      _$FoodCategoryFromJson(json);
}

@freezed
abstract class FoodItem with _$FoodItem {
  const factory FoodItem({
    required int id,
    @JsonKey(name: 'category_id') required int categoryId,
    required String key,
    @JsonKey(name: 'label_tr') required String labelTr,
    @JsonKey(name: 'label_en') required String labelEn,
    @JsonKey(name: 'is_custom') @Default(false) bool isCustom,
  }) = _FoodItem;

  factory FoodItem.fromJson(Map<String, dynamic> json) =>
      _$FoodItemFromJson(json);
}

File: backend/internal/models/venue.go
package models

import "time"

type VenueStatus string

const (
	VenueStatusPending   VenueStatus = "pending"
	VenueStatusApproved  VenueStatus = "approved"
	VenueStatusRejected  VenueStatus = "rejected"
	VenueStatusSuspended VenueStatus = "suspended"
)

type TrustCriteria struct {
	ID          int     `json:"id"`
	Key         string  `json:"key"`
	Name        string  `json:"name"`
	Description *string `json:"description,omitempty"`
}

type FoodCategory struct {
	ID       int     `json:"id"`
	Key      string  `json:"key"`
	Name     string  `json:"name"`
	ImageURL *string `json:"image_url,omitempty"`
}

type VenuePhoto struct {
	ID         string    `json:"id"`
	VenueID    string    `json:"venue_id"`
	URL        string    `json:"url"`
	UploadedBy string    `json:"uploaded_by"`
	IsPrimary  bool      `json:"is_primary"`
	CreatedAt  time.Time `json:"created_at"`
}

type Venue struct {
	ID               string          `json:"id"`
	Name             string          `json:"name"`
	City             string          `json:"city"`
	District         *string         `json:"district,omitempty"`
	Latitude         float64         `json:"latitude"`
	Longitude        float64         `json:"longitude"`
	GooglePlaceID    *string         `json:"google_place_id,omitempty"`
	Notes            *string         `json:"notes"`
	Status           VenueStatus     `json:"status"`
	RejectionNote    *string         `json:"rejection_note,omitempty"`
	AddedBy          string          `json:"added_by"`
	AddedByName      *string         `json:"added_by_name,omitempty"`
	AddedBySurname   *string         `json:"added_by_surname,omitempty"`
	AddedByEmail     *string         `json:"added_by_email,omitempty"`
	ApprovedBy       *string         `json:"approved_by,omitempty"`
	VerifiedAt          *time.Time      `json:"verified_at,omitempty"`
	VerificationDueAt   *time.Time      `json:"verification_due_at,omitempty"`
	LastNotifiedAt      *time.Time      `json:"last_notified_at,omitempty"`
	Distance            *float64        `json:"distance,omitempty"` // metre cinsinden, yakın mekan sorgusunda dolar
	FoodHalalMode    string          `json:"food_halal_mode"`
	ExcludedProducts []string        `json:"excluded_products"`
	TrustCriteria    []TrustCriteria `json:"trust_criteria"`
	Photos           []VenuePhoto    `json:"photos"`
	Categories       []FoodCategory  `json:"categories"`
	AverageRating    *float64        `json:"average_rating,omitempty"`
	ReviewCount      int             `json:"review_count"`
	ConfirmationCount int             `json:"confirmation_count"`
	Badge            *Badge          `json:"badge,omitempty"`
	// ConfirmedByMe — yalnızca giriş yapmış viewer için detay sorgusunda doldurulur:
	// bu kullanıcı mekanı içinde bulunulan dönemde doğrulamış mı? Anonim/diğer
	// sorgularda nil kalır (omitempty ile JSON'a yazılmaz).
	ConfirmedByMe    *bool           `json:"confirmed_by_me,omitempty"`
	// ConfirmedAt — yalnızca "doğruladığım mekanlar" listesinde dolar: bu
	// rehberin mekanı ne zaman doğruladığı. Diğer sorgularda nil kalır.
	ConfirmedAt      *time.Time      `json:"confirmed_at,omitempty"`
	CategoriesStr    *string         `json:"categories_str,omitempty"`
	CreatedAt        time.Time       `json:"created_at"`
	UpdatedAt        time.Time       `json:"updated_at"`
}

// Badge — mekanın dönemsel taze onay sayısından türetilen güven rozeti.
type Badge struct {
	Level string `json:"level"` // base | bronze | silver | gold | platinum
	Count int    `json:"count"` // son periyottaki farklı doğrulayan sayısı (ekleyen dahil)
}

// BadgeFromCount — dönemsel onay sayısını rozet seviyesine çevirir.
func BadgeFromCount(count int) Badge {
	level := "base"
	switch {
	case count >= 11:
		level = "platinum"
	case count >= 6:
		level = "gold"
	case count >= 2:
		level = "silver"
	case count >= 1:
		level = "bronze"
	}
	return Badge{Level: level, Count: count}
}
